src/database.ts

```diff
@@ -55,24 +65,110 @@ import {
   TransactionRunner,
 } from './transaction-runner';
 
-export interface GetDatabaseOptions {
+import {google as spannerClient} from '../proto/spanner';
+import {Schema, BasicResponse} from './common';
+import {ServiceError, CallOptions} from 'grpc';
+import {Readable, Transform} from 'stream';
+import {Stats} from 'fs';
+import {StreamProxy} from 'google-gax/build/src/streamingCalls/streaming';
+
+interface CreateBatchTransactionCallback {
+  (
+    err?: ServiceError | null,
+    transaction?: BatchTransaction | null,
+    apiResponse?:
+      | spannerClient.spanner.v1.ITransaction
+      | spannerClient.spanner.v1.ISession
+  ): void;
+}
+
+type CreateBatchTransactionResponse = [
+  BatchTransaction,
+  spannerClient.spanner.v1.ITransaction | spannerClient.spanner.v1.ISession
+];
+
+interface GetDatabaseOptions {
   autoCreate?: boolean;
 }
-export type DatabaseResponse = [Database, r.Response];
-export interface DatabaseCallback {
+type DatabaseResponse = [Database, r.Response];
+interface DatabaseCallback {
   (err: Error | null, database?: Database, apiResponse?: r.Response): void;
 }
 
-export interface GetSnapshotCallback {
+interface GetSnapshotCallback {
   (err: Error, snapshot?: null): void;
   (err: null, snapshot: Snapshot): void;
 }
 
-export interface GetTransactionCallback {
-  (err: Error, transaction?: null): void;
-  (err: null, transaction: Transaction): void;
+interface GetTransactionCallback {
+  (err?: Error | null, transaction?: Transaction | null): void;
+}
+
+interface SessionPoolCtor {
+  (database: Database, option: SessionPoolOptions | null): void;
+}
+
+interface UpdateSchemaCallback {
+  (
+    err: Error | null,
+    operation: GaxOperation,
+    resp: databaseAdmin.longrunning.IOperation
+  ): void;
+}
+
+interface PoolRequestConfig {
+  reqOpts: {
+    // tslint:disable-next-line: no-any
+    [key: string]: any;
+    session?: string;
+  };
+  client?: string;
+  method?: string;
+  gaxOptions: CallOptions;
+}
+
+interface PoolRequestCallback {
+  (err: Error | null, ...args: Array<{} | null>): void;
 }
 
+interface RunCallback {
+  (err: Error | null, rows: Row[]): Stats;
+}
+
+type GetSessionsOptions = spannerClient.spanner.v1.IListSessionsRequest &
+  CallOptions;
+
+type GetMetadataResponse = [databaseAdmin.spanner.admin.database.v1.IDatabase];
+interface GetMetadataCallback {
+  (
+    err: Error | null,
+    metadata?: databaseAdmin.spanner.admin.database.v1.IDatabase
+  ): void;
+}
+
+interface GetSchemaCallback {
+  (err: Error | null, ...args: Array<{} | string>): void;
```

Comment:
This doesn't really provide any type safety, I think we should mirror what we did in [PubSub](https://github.com/googleapis/nodejs-pubsub/blob/master/src/pubsub.ts#L148)